tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ImdsV2Tests.cs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.AppConfig;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.ManagedIdentity;
+using Microsoft.Identity.Test.Common.Core.Mocks;
+using Microsoft.Identity.Test.Unit.Helpers;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Identity.Test.Unit.ManagedIdentityTests
+{
+    [TestClass]
+    public class ImdsV2Tests : TestBase
+    {
+        private const string serverHeaderImdsVersion = "IMDS/150.870.65.1325";
+        private const string CsrMetadataEndpoint = "http://169.254.169.254/metadata/identity/getPlatformMetadata";
+        private readonly TestRetryPolicyFactory _testRetryPolicyFactory = new TestRetryPolicyFactory();
+
+        [TestMethod]
+        public async Task GetCsrMetadataAsyncSucceeds()
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                var miBuilder = ManagedIdentityApplicationBuilder.Create(ManagedIdentityId.SystemAssigned)
+                    .WithHttpManager(httpManager)
+                    .WithRetryPolicyFactory(_testRetryPolicyFactory);
+                var managedIdentityApp = miBuilder.BuildConcrete();
+                RequestContext requestContext = new RequestContext(managedIdentityApp.ServiceBundle, Guid.NewGuid(), null);
+
+                httpManager.AddManagedIdentityMockHandler(
+                    CsrMetadataEndpoint,
+                    ManagedIdentityTests.Resource,
+                    MockHelpers.GetCsrMetadataSuccessfulResponse(),
+                    ManagedIdentitySource.ImdsV2,
+                    serverHeaderImdsVersion: serverHeaderImdsVersion);
+
+                bool isValid = await ImdsV2ManagedIdentitySource.GetCsrMetadataAsync(requestContext).ConfigureAwait(false);
```

Comment:
Don't we expose the source in the public API @gladjohn ?

Reply:
Yes, we do. Why are you asking? You want me to call that api here?

Reply:
addressed in other comment